### PR TITLE
Internal to External Sync - 10/06/2016

### PR DESCRIPTION
**Pull request by tirodkar**
What's new in this week's sync for EarlGrey?

* A better way to blacklist URL's in GREYConfiguration.
* A Stopwatch class that times EarlGrey interactions and drains.
* A verbose logger that provides more info for what's going on in your EarlGrey tests.
* A better way to print the hierarchy for debugging.
* Improvements to our error handling....

### CODE REVIEW

File: EarlGrey/Common/GREYTestCaseInvocation.h [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+//
+
+#import <Foundation/Foundation.h>
+
+/**
+ *  Class used for invoking a test case method and updating the status of the test case
+ */
+@interface GREYTestCaseInvocation : NSInvocation
+
+@end
```
Comment (sid-github) [inline EarlGrey/Common/GREYTestCaseInvocation.h]: missing nl.

**Comment by sid-github** [comment]
LGTM

File: EarlGrey/Common/GREYTestCaseInvocation.h (hunk 1 above)
Comment (tirodkar) [inline EarlGrey/Common/GREYTestCaseInvocation.h]: Done

File: EarlGrey/Additions/NSObject+GREYAdditions.m [hunk 2]
```diff
@@ -60,24 +66,36 @@ + (void)load {
...
+  if ([self grey_isWebAccessibilityElement]) {
+    return [GREYElementHierarchy hierarchyStringForElement:[self grey_viewContainingSelf]];
+  } else if ([self isKindOfClass:[UIView class]] ||
+             [self respondsToSelector:@selector(accessibilityContainer)]) {
+    return [GREYElementHierarchy hierarchyStringForElement:self];
+  } else {
+    NSAssert(NO, @"The element hierarchy call is being made on an element that is not a valid "
+                 @"UI element.");
+  }
+  return nil;
```
Comment (khandpur) [inline EarlGrey/Additions/NSObject+GREYAdditions.m]: make this part of the else statement?

File: EarlGrey/Common/GREYElementHierarchy.h [hunk 3]
```diff
@@ -44,4 +44,9 @@
  */
 + (NSString *)hierarchyStringForElement:(id)element
                withAnnotationDictionary:(NSDictionary *)annotationDictionary;
+
+/**
+ *  Returns the UI hierarchy for all UI windows provided by the GREYUIWindowProvider.
```
Comment (khandpur) [inline EarlGrey/Common/GREYElementHierarchy.h]: @c UIWindows

File: EarlGrey/Common/GREYVisibilityChecker.m [hunk 4]
```diff
@@ -22,6 +22,7 @@
 #import "Additions/NSObject+GREYAdditions.h"
 #import "Additions/UIView+GREYAdditions.h"
 #import "Common/GREYConstants.h"
+#import "Common/GREYVerboseLogger.h"
```
Comment (khandpur) [inline EarlGrey/Common/GREYVisibilityChecker.m]: maintain sort order

File: EarlGrey/Core/GREYElementInteraction.m [hunk 5]
```diff
@@ -24,6 +24,8 @@
 #import "Assertion/GREYAssertions.h"
 #import "Common/GREYConfiguration.h"
 #import "Common/GREYDefines.h"
+#import "Common/GREYStopwatch.h"
+#import "Common/GREYVerboseLogger.h"
```
Comment (khandpur) [inline EarlGrey/Core/GREYElementInteraction.m]: sort order

File: EarlGrey/Common/GREYElementHierarchy.h (hunk 3 above)
Comment (tirodkar) [inline EarlGrey/Common/GREYElementHierarchy.h]: Done.

File: EarlGrey/Common/GREYVisibilityChecker.m (hunk 4 above)
Comment (tirodkar) [inline EarlGrey/Common/GREYVisibilityChecker.m]: Done.

File: EarlGrey/Core/GREYElementInteraction.m (hunk 5 above)
Comment (tirodkar) [inline EarlGrey/Core/GREYElementInteraction.m]: Done.

**Comment by tirodkar** [comment]
@khandpur, Updated.

**Comment by sid-github** [comment]
LGTM

**Pull request merged**